node/bft/src/helpers/pending.rs

```diff
@@ -128,12 +131,29 @@ impl<T: Copy + Clone + PartialEq + Eq + Hash, V: Clone> Pending<T, V> {
 
     /// Removes the callbacks for the specified `item` that have expired.
     pub fn clear_expired_callbacks_for_item(&self, item: impl Into<T>) {
+        let item = item.into();
+        // Acquire the callbacks lock.
+        let mut callbacks = self.callbacks.lock();
         // Clear the callbacks that have expired.
-        if let Some(callbacks) = self.callbacks.lock().get_mut(&item.into()) {
+        if let Some(callback_values) = callbacks.get_mut(&item) {
             // Fetch the current timestamp.
             let now = OffsetDateTime::now_utc().unix_timestamp();
             // Remove the callbacks that have expired.
-            callbacks.retain(|(_, timestamp)| now - *timestamp <= CALLBACK_TIMEOUT_IN_SECS);
+            callback_values.retain(|(_, timestamp)| now - *timestamp <= CALLBACK_EXPIRATION_IN_SECS);
+
+            // If there are no more remaining callbacks for the item, remove the item from the pending queue.
+            if callback_values.is_empty() {
+                callbacks.remove(&item);
+                self.pending.write().remove(&item);
+            }
+        }
+    }
+
+    /// Removes the callbacks for all items have that expired.
+    pub fn clear_expired_callbacks(&self) {
+        let items = self.pending.read().keys().copied().collect::<Vec<T>>();
+        for item in items.into_iter() {
+            self.clear_expired_callbacks_for_item(item);
```

Comment:
I missed this earlier, but if we can expect many items here, it might be better to obtain the current timestamp and the lock over `callbacks` before this loop